src/passes/OptimizeInstructions.cpp

```diff
@@ -1835,17 +1835,46 @@ struct OptimizeInstructions
         Bits::getMaxBits(curr->ifFalse, this) <= 1) {
       // The condition and both arms are i32 booleans, which allows us to do
       // boolean optimizations.
-      Expression* x;
-      Expression* y;
 
       // x ? y : 0   ==>   x & y
-      if (matches(curr, select(any(&y), ival(0), any(&x)))) {
-        return builder.makeBinary(AndInt32, y, x);
+      if (matches(curr->ifFalse, ival(0))) {
+        return builder.makeBinary(AndInt32, curr->ifTrue, curr->condition);
       }
 
       // x ? 1 : y   ==>   x | y
-      if (matches(curr, select(ival(1), any(&y), any(&x)))) {
-        return builder.makeBinary(OrInt32, y, x);
+      if (matches(curr->ifTrue, ival(1))) {
+        return builder.makeBinary(OrInt32, curr->ifFalse, curr->condition);
+      }
+
+      // Some operations require us to add an eqz, like
+      // x ? 0 : y   ==>   !x & y
+      // We replace the select with an and, which does not change code size, and
+      // then reduce code size by removing the number (2 bytes) and adding an
+      // eqz (one byte). However, the eqz adds work, so it is not obvious that
+      // this is beneficial to do - only do it when we know we can fold away the
+      // eqz.
+      // TODO: Perhaps also do this even if we can't remove the eqz, if we are
+      //       optimizing for size? We should measure the performance aspect
+      //       first though, to see if we need that.
+      //
+      // (Note that aside from a relational binary which we can flip, we could
+      // also check for an eqz unary, but a select with an eqz condition would
+      // have already been removed in the cases relevant to us, since the true
+      // and false arms can be flipped given that one of them is a constant.)
+      if (auto* binary = curr->condition->dynCast<Binary>()) {
+        if (binary->isRelational()) {
+          // x ? 0 : y   ==>   !x & y
+          if (matches(curr->ifTrue, ival(0))) {
+            binary->op = invertBinaryOp(binary->op);
```

Comment:
Also note `invertBinaryOp` should check for `InvalidBinary` which can be appear for some relational floating point ops like `<`, `>` `>=`, `<=` due to NaN behavior

Reply:
True, but we have checked above that all the children of the select are `i32` here, so that should not be a problem.

Reply:
But it's not relate to `ifFalse` or `ifTrue`. `condition` also have `i32` result while binary operation of `condition` could be floating point. I mean this case:
```wat
x > 2.0 ? 1 : y < 0.0
```
where `x` and `y` are `f64` but all expressions results are `i32`.

Reply:
I see, thanks. Ok, then this PR needs more work then.

Reply:
if just need to add one more simple check:
```cpp
BinaryOp op  = invertBinaryOp(binary->op);
if (op != InvalidBinary) {
  binary->op =  op;
  return builder.makeBinary(AndInt32, curr->ifFalse, binary);
}
```
and same for second case